spacy/tests/doc/test_json_doc_conversion.py

```diff
@@ -57,18 +102,104 @@ def test_doc_to_json(doc):
 
 
 def test_doc_to_json_underscore(doc):
     Doc.set_extension("json_test1", default=False)
     Doc.set_extension("json_test2", default=False)
     doc._.json_test1 = "hello world"
     doc._.json_test2 = [1, 2, 3]
+
     json_doc = doc.to_json(underscore=["json_test1", "json_test2"])
     assert "_" in json_doc
     assert json_doc["_"]["json_test1"] == "hello world"
     assert json_doc["_"]["json_test2"] == [1, 2, 3]
     assert not schemas.validate(schemas.DocJSONSchema, json_doc)
+    assert srsly.json_loads(srsly.json_dumps(json_doc)) == json_doc
+
+
+def test_doc_to_json_with_token_span_attributes(doc):
+    Doc.set_extension("json_test1", default=False)
+    Doc.set_extension("json_test2", default=False)
+    Token.set_extension("token_test", default=False)
+    Span.set_extension("span_test", default=False)
+
+    doc._.json_test1 = "hello world"
+    doc._.json_test2 = [1, 2, 3]
+    doc[0:1]._.span_test = "span_attribute"
+    doc[0]._.token_test = 117
+    doc.spans["span_group"] = [doc[0:1]]
+    json_doc = doc.to_json(
+        underscore=["json_test1", "json_test2", "token_test", "span_test"]
+    )
+
+    assert "_" in json_doc
+    assert json_doc["_"]["json_test1"] == "hello world"
+    assert json_doc["_"]["json_test2"] == [1, 2, 3]
+    assert "underscore_token" in json_doc
+    assert "underscore_span" in json_doc
+    assert json_doc["underscore_token"]["token_test"]["value"] == 117
+    assert json_doc["underscore_span"]["span_test"]["value"] == "span_attribute"
+    assert not schemas.validate(schemas.DocJSONSchema, json_doc)
+    assert srsly.json_loads(srsly.json_dumps(json_doc)) == json_doc
+
+
+def test_doc_to_json_with_custom_user_data(doc):
+    Doc.set_extension("json_test", default=False)
+    Token.set_extension("token_test", default=False)
+    Span.set_extension("span_test", default=False)
+
+    doc._.json_test = "hello world"
+    doc[0:1]._.span_test = "span_attribute"
+    doc[0]._.token_test = 117
+    json_doc = doc.to_json(underscore=["json_test", "token_test", "span_test"])
+    doc.user_data["user_data_test"] = 10
+    doc.user_data[("user_data_test2", True)] = 10
+
+    assert "_" in json_doc
+    assert json_doc["_"]["json_test"] == "hello world"
+    assert "underscore_token" in json_doc
+    assert "underscore_span" in json_doc
+    assert json_doc["underscore_token"]["token_test"]["value"] == 117
+    assert json_doc["underscore_span"]["span_test"]["value"] == "span_attribute"
+    assert not schemas.validate(schemas.DocJSONSchema, json_doc)
+    assert srsly.json_loads(srsly.json_dumps(json_doc)) == json_doc
+
+
+def test_doc_to_json_with_token_span_same_identifier(doc):
+    Doc.set_extension("my_ext", default=False)
+    Token.set_extension("my_ext", default=False)
+    Span.set_extension("my_ext", default=False)
+
+    doc._.my_ext = "hello world"
+    doc[0:1]._.my_ext = "span_attribute"
+    doc[0]._.my_ext = 117
+    json_doc = doc.to_json(underscore=["my_ext"])
```

Comment:
This will fail until the schema is updated:

```suggestion
    json_doc = doc.to_json(underscore=["my_ext"])
    
    assert schemas.validate(schemas.DocJSONSchema, json_doc)
```

Reply:
I looked into the `validate` command, [which returns an empty `list` if the schema is correct](https://github.com/explosion/spaCy/blob/master/spacy/schemas.py#L42). However, `assert []` causes an `assertion error` even though the schema is correct. 

I think the assertion test should be:
`assert not schemas.validate(schemas.DocJSONSchema, json_doc)` or 
`assert schemas.validate(schemas.DocJSONSchema, json_doc) == []`

Reply:
Ah, I misunderstood how the validation works, and I also thought that extra fields were forbidden when they aren't. I'm not sure we want to make extra fields forbidden, but based on trying things out it turns out that `_` causes problems in the schema, so maybe this would be better:

```
    underscore_doc: Optional[Dict[StrictStr, Any]] = Field(
        None, title="Any custom data stored in the document's _ attribute",
        alias="_",
    )
```

Reply:
Updated the code 👍

Reply:
Can you still add the right version of this validation to the test?

Reply:
I'm also a little concerned that the schema makes it look like it supports `underscore_doc` when it doesn't, but otherwise the schema kind of doesn't work properly. And we do the same thing in `TokenPattern`, too.

Reply:
What do you mean by "right version"? 😇

Reply:
Whatever works at this line above if the validation passes.

Reply:
Do you mean this validation line?
`assert schemas.validate(schemas.DocJSONSchema, json_doc)`

Reply:
And I think we should change all the validation lines in the tests because this is too confusing:

```python
assert not schemas.validate()
```

Maybe:

```python
assert len(schemas.validate()) == 0
```